### PR TITLE
Automatic SteamVR tracker switching

### PR DESCRIPTION
**Pull request by ImUrX**
Added a detection of the trackers assigned to the body and it switches the enabled SteamVR trackers depending on it.

Also added a hands warning modal because god, we cant have enough warnings

Requires SlimeVR/SolarXR-Protocol#132

### CODE REVIEW

File: gui/public/i18n/en/translation.ftl [hunk 1]
```diff
@@ -283,6 +283,13 @@ settings-general-steamvr-trackers-feet = Feet
 settings-general-steamvr-trackers-knees = Knees
 settings-general-steamvr-trackers-elbows = Elbows
 settings-general-steamvr-trackers-hands = Hands
+settings-general-steamvr-trackers-tracker_toggling = Automatic tracker switching
+settings-general-steamvr-trackers-tracker_toggling-description = Automatically handles the turning on and off of SteamVR trackers depending on your current body assignment of trackers
+settings-general-steamvr-trackers-tracker_toggling-label = Automatic tracker switching
+settings-general-steamvr-trackers-hands-warning = <b>Warning:</b> Hands trackers will replace the tracking
+    of your controllers, are you sure?
```
Comment (Louka3000) [inline gui/public/i18n/en/translation.ftl:290]:
```suggestion
settings-general-steamvr-trackers-tracker_toggling = Automatic tracker assignment
settings-general-steamvr-trackers-tracker_toggling-description = Automatically handles toggling SteamVR trackers on or off depending on your current tracker body assignments
settings-general-steamvr-trackers-tracker_toggling-label = Automatic tracker assignment
settings-general-steamvr-trackers-hands-warning = <b>Warning:</b> hands trackers will override your controllers. Are you sure?
```

Not too sure about "current tracker body assignments", but it sounds better than "current body assignment of trackers" to me.

**Review by Louka3000 — CHANGES_REQUESTED** [review]

File: gui/public/i18n/en/translation.ftl (hunk 1 above)
Comment (ImUrX) [reply gui/public/i18n/en/translation.ftl:290]: seems good, the line break is in purpose as the warning becomes *loong* otherwise

Comment (Louka3000) [reply gui/public/i18n/en/translation.ftl:290]: Added a line break before “Are you sure?”. How’s that?

Comment (ImUrX) [reply gui/public/i18n/en/translation.ftl:290]:
```suggestion
settings-general-steamvr-trackers-tracker_toggling = Automatic tracker assignment
settings-general-steamvr-trackers-tracker_toggling-description = Automatically handles toggling SteamVR trackers on or off depending on your current tracker assignments
settings-general-steamvr-trackers-tracker_toggling-label = Automatic tracker assignment
settings-general-steamvr-trackers-hands-warning = <b>Warning:</b> hands trackers will override your controllers.
    Are you sure?
```
this fine?

File: server/desktop/src/main/java/dev/slimevr/desktop/platform/SteamVRBridge.kt [hunk 2]
```diff
@@ -0,0 +1,425 @@
...
+
+		// hasFeet if lower and/or upper leg tracker is on
+		val hasFeet = instance.humanPoseManager.skeleton.let {
+			(it.leftUpperLegTracker != null || it.leftLowerLegTracker != null) &&
+				(it.rightUpperLegTracker != null || it.rightLowerLegTracker != null)
+		}
+		changeShareSettings(TrackerRole.LEFT_FOOT, hasFeet)
+		changeShareSettings(TrackerRole.RIGHT_FOOT, hasFeet)
+
+		// hasKnees if feet trackers are on (maybe check for at least 1 leg tracker?)
```
Comment (Louka3000) [inline server/desktop/src/main/java/dev/slimevr/desktop/platform/SteamVRBridge.kt:74]:
```suggestion
		// hasKnees if foot tracker and lower and/or upper leg tracker is on
```
damn, just copy-pasted from Discord DMs without checking smh

**Review by Louka3000 — APPROVED** [review]
lgtm

File: server/desktop/src/main/java/dev/slimevr/desktop/platform/SteamVRBridge.kt (hunk 2 above)
Comment (ImUrX) [reply server/desktop/src/main/java/dev/slimevr/desktop/platform/SteamVRBridge.kt:74]: yes

**Review by Louka3000 — APPROVED** [review]

File: server/core/src/main/java/dev/slimevr/protocol/rpc/settings/RPCSettingsHandler.kt [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+						.setValue(
+							SkeletonConfigValues.KNEE_TRACKER_ANKLE_AVERAGING,
+							max(0.0, ratios.interpKneeTrackerAnkle().toDouble()).toFloat(),
+						)
+				}
+				if (ratios.hasInterpKneeAnkle()) {
+					hpm
+						.setValue(
+							SkeletonConfigValues.KNEE_ANKLE_AVERAGING,
+							max(0.0, ratios.interpKneeAnkle().toDouble()).toFloat(),
```
Comment (ButterscotchV) [inline server/core/src/main/java/dev/slimevr/protocol/rpc/settings/RPCSettingsHandler.kt:309]: wut is this
pls use float `max` or `coerceAtLeast`
https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.ranges/coerce-at-least.html
```suggestion
							max(0f, ratios.imputeWaistFromChestHip()),
						)
				}
				if (ratios.hasImputeWaistFromChestLegs()) {
					hpm
						.setValue(
							SkeletonConfigValues.WAIST_FROM_CHEST_LEGS_AVERAGING,
							max(0f, ratios.imputeWaistFromChestLegs()),
						)
				}
				if (ratios.hasImputeHipFromChestLegs()) {
					hpm
						.setValue(
							SkeletonConfigValues.HIP_FROM_CHEST_LEGS_AVERAGING,
							max(0f, ratios.imputeHipFromChestLegs()),
						)
				}
				if (ratios.hasImputeHipFromWaistLegs()) {
					hpm
						.setValue(
							SkeletonConfigValues.HIP_FROM_WAIST_LEGS_AVERAGING,
							max(0f, ratios.imputeHipFromWaistLegs()),
						)
				}
				if (ratios.hasInterpHipLegs()) {
					hpm
						.setValue(
							SkeletonConfigValues.HIP_LEGS_AVERAGING,
							max(0f, ratios.interpHipLegs()),
						)
				}
				if (ratios.hasInterpKneeTrackerAnkle()) {
					hpm
						.setValue(
							SkeletonConfigValues.KNEE_TRACKER_ANKLE_AVERAGING,
							max(0f, ratios.interpKneeTrackerAnkle()),
						)
				}
				if (ratios.hasInterpKneeAnkle()) {
					hpm
						.setValue(
							SkeletonConfigValues.KNEE_ANKLE_AVERAGING,
							max(0f, ratios.interpKneeAnkle()),
```

File: server/core/src/main/java/dev/slimevr/protocol/rpc/settings/RPCSettingsHandler.kt [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+
+			RPCSettingsBuilder.readAutoBoneSettings(autoBoneSettings, autoBoneConfig)
+		}
+
+		if (req.resetsSettings() != null) {
+			val resetsConfig = api.server.configManager
+				.vrConfig
+				.resetsConfig
+			val mode = ArmsResetModes
+				.fromId(max(req.resetsSettings().armsMountingResetMode().toDouble(), 0.0).toInt())
```
Comment (ButterscotchV) [inline server/core/src/main/java/dev/slimevr/protocol/rpc/settings/RPCSettingsHandler.kt:338]: 😿pls use int `max` or `coerceAtLeast`
https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.ranges/coerce-at-least.html
```suggestion
				.fromId(max(req.resetsSettings().armsMountingResetMode(), 0))
```

File: server/desktop/src/main/java/dev/slimevr/desktop/platform/SteamVRBridge.kt [hunk 5]
```diff
@@ -0,0 +1,426 @@
...
+
+		// batteryTrackers is filled with trackers that would give battery data
+		// for the SteamVR tracker according to its role. Warning: trackers
+		// inside batteryTrackers could be null, so there must be a null check
+		// when accessing its data.
+		batteryTrackers
+			.add(
+				getTrackerForSkeleton(
+					allTrackers,
+					getByTrackerRole(role) ?: return,
```
Comment (ButterscotchV) [inline server/desktop/src/main/java/dev/slimevr/desktop/platform/SteamVRBridge.kt:191]: I feel like I asked this before and there was a good reason, but why don't we just use `localTracker.trackerPosition` here?

File: server/desktop/src/main/java/dev/slimevr/desktop/platform/SteamVRBridge.kt [hunk 6]
```diff
@@ -0,0 +1,426 @@
...
+		changeShareSettings(TrackerRole.RIGHT_KNEE, hasKnees)
+
+		// hasElbows if an upper arm or a lower arm tracker is on
+		val hasElbows = instance.humanPoseManager.skeleton.let { it.hasLeftArmTracker && it.hasRightArmTracker }
+		changeShareSettings(TrackerRole.LEFT_ELBOW, hasElbows)
+		changeShareSettings(TrackerRole.RIGHT_ELBOW, hasElbows)
+
+		// Hands aren't touched as they will override the controller's tracking
+		return true
+	}
```
Comment (ButterscotchV) [inline server/desktop/src/main/java/dev/slimevr/desktop/platform/SteamVRBridge.kt:88]: Why not just keep `instance.humanPoseManager.skeleton` in a local variable? It's so long and you use it a ton here.

**Review by ButterscotchV — CHANGES_REQUESTED** [review]

File: server/desktop/src/main/java/dev/slimevr/desktop/platform/SteamVRBridge.kt (hunk 6 above)
Comment (ImUrX) [reply server/desktop/src/main/java/dev/slimevr/desktop/platform/SteamVRBridge.kt:88]: yeah we could, we no longer reinitialize skeleton anymore, right?

File: server/desktop/src/main/java/dev/slimevr/desktop/platform/SteamVRBridge.kt (hunk 5 above)
Comment (ImUrX) [reply server/desktop/src/main/java/dev/slimevr/desktop/platform/SteamVRBridge.kt:191]: i sadly dont remember, and cant seem to find the reason why qwq

Comment (ButterscotchV) [reply server/desktop/src/main/java/dev/slimevr/desktop/platform/SteamVRBridge.kt:188]: Well uhhhhh I guess it's fine to just leave it that way lol, maybe put a comment to look into it or smth

File: server/desktop/src/main/java/dev/slimevr/desktop/platform/SteamVRBridge.kt [hunk 7]
```diff
@@ -0,0 +1,423 @@
...
+			}
+		}
+		return false
+	}
+
+	override fun updateShareSettingsAutomatically(): Boolean {
+		if (!config.automaticSharedTrackersToggling) return false
+		// Enable waist if skeleton has an spine tracker
+		changeShareSettings(TrackerRole.WAIST, instance.humanPoseManager.skeleton.hasSpineTracker)
+		val skeleton = instance.humanPoseManager.skeleton
```
Comment (ButterscotchV) [inline server/desktop/src/main/java/dev/slimevr/desktop/platform/SteamVRBridge.kt:59]: Why not use the variable here lol
```suggestion
		val skeleton = instance.humanPoseManager.skeleton
		// Enable waist if skeleton has an spine tracker
		changeShareSettings(TrackerRole.WAIST, skeleton.hasSpineTracker)
```

**Review by ButterscotchV — APPROVED** [review]

**Pull request merged**